File: 78-subsets/SolutionRecursive.java
public class Solution {
    void bktk(int[] nums, int lev, ArrayList<Integer> cur, List<List<Integer>> result){
        result.add(new ArrayList<Integer>(cur));
        for(int i=lev; i<nums.length; i++){
            cur.add(nums[i]);
            bktk(nums, i+1, cur, result);
            cur.remove(cur.size()-1);
        }
    }
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> result = new ArrayList<List<Integer>>();
        ArrayList<Integer> cur =  new ArrayList<Integer>();
        bktk(nums, 0, cur, result);
        return result;
    }
}

File: Leetcode/78-subsets/SolutionRecursive.cc
class Solution {
public:
    void bktk(const vector<int>& a, int lev, vector<int>& cur, vector<vector<int>>& r){
        r.push_back(cur);
        for(int i=lev; i<a.size(); i++){
            cur.push_back(a[i]);
            bktk(a, i+1, cur, r);
            cur.pop_back();
        }
    }
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> r;
        vector<int> cur;
        bktk(nums, 0, cur, r);
        return r;
    }
};